docs/SIL.rst

```diff
@@ -2422,19 +2422,17 @@ A few sorts of SIL value have lifetimes that are constrained that way:
 
 1: `begin_borrow [lexical]`
 2: `move_value [lexical]`
-3: @owned function arguments
+3: @owned/@guaranteed function arguments
 4: `alloc_stack [lexical]`
 
-That these three have constrained lifetimes is encoded in ValueBase::isLexical,
-which should be checked before changing the lifetime of a value.
+That the first three have constrained lifetimes is encoded in
+ValueBase::isLexical, which should be checked before changing the lifetime of a
+value.
 
-The reason that only @owned function arguments are constrained is that a
-@guaranteed function argument is guaranteed by the function's caller to live for
-the full duration of the function already.  Optimization of the function alone
-can't shorten it.  When such a function is inlined into its caller, though, a
-lexical borrow scope is added for each of its @guaranteed arguments, ensuring
-that the lifetime of the corresponding source-level value is not shortened in a
-way that doesn't respect deinit barriers.
+While both @guaranteed and @owned function arguments are lexical, this is only
+interesting in the case of @owned arguments: @guaranteed function arguments are
+already guaranteed by the function's caller to live for the full duration of
+the function already.  Optimization of the function alone can't shorten it.  
```

Comment:
At the language level, arguments are lexical under the same conditions as variables. Namely, that they don't have eager-move semantics, which includes CoW types and no-implicit-copy type qualifiers.

Note that when a user explicitly writes `borrow` or `consume`, the arguments are not lexical. Whether an argument is lexical is independent of the ownership convention.

Reply:
Rewrote this.